packages/block-editor/src/hooks/border-color.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import ColorGradientControl from '../components/colors-gradients/control';
+import {
+	getColorClassName,
+	getColorObjectByColorValue,
+} from '../components/colors';
+import useEditorFeature from '../components/use-editor-feature';
+import { hasBorderFeatureSupport, shouldSkipSerialization } from './border';
+import { cleanEmptyObject } from './utils';
+
+// Defining empty array here instead of inline avoids unnecessary re-renders of
+// color control.
+const EMPTY_ARRAY = [];
+
+/**
+ * Inspector control panel containing the border color related configuration.
+ *
+ * There is deliberate overlap between the colors and borders block supports
+ * relating to border color. It can be argued the border color controls could
+ * be included within either, or both, the colors and borders panels in the
+ * inspector controls. If they share the same block attributes it should not
+ * matter.
+ *
+ * @param  {Object} props Block properties.
+ * @return {WPElement}    Border color edit element.
+ */
+export function BorderColorEdit( props ) {
+	const {
+		attributes: { borderColor, style },
+		setAttributes,
+	} = props;
+	const colors = useEditorFeature( 'color.palette' ) || EMPTY_ARRAY;
+
+	const disableCustomColors = ! useEditorFeature( 'color.custom' );
+	const disableCustomGradients = ! useEditorFeature( 'color.customGradient' );
+
+	if ( useIsBorderColorDisabled( props ) ) {
+		return null;
+	}
+
+	const onChangeColor = ( value ) => {
+		const colorObject = getColorObjectByColorValue( colors, value );
+		const newStyle = {
+			...style,
+			border: {
+				...style?.border,
+				color: colorObject?.slug ? undefined : value,
```

Comment:
I'm just testing this PR and, for me, the border color panel doesn't add the style to the root element. The `has-{color}-border-color` class is present.

If I change this line to `color: colorObject?.color || value,` it works as I'd expect.

![color-undefined](https://user-images.githubusercontent.com/6458278/114345016-4638c700-9ba4-11eb-99bf-8f227064ef13.gif)

Is there a reason why the style prop should be `undefined` here? 

Or are there CSS definitions for `.has-{color}-border-color` classes that I'm not seeing (or something else I'm missing)?

Thanks!

Reply:
There's a good chance I missed something here so I'll look into it. Thanks for testing.

> Is there a reason why the style prop should be undefined here?

This was `undefined` so that when using a named color it uses the CSS classes generated for the color from the current palette. If it stored the color value there the inline style overrules it. This could become a problem if the theme changes the color palette or the user switches themes.

Reply:
> Strangely, changing the text color modifies the border color, though I can't see why in the computed styles.

I have the suspicion that the theme sets the border color to the `currentColor` which would be determined by the text color. The inline border style works for me when a border style is set though.

Reply:
> This was undefined so that when using a named color it uses the CSS classes generated for the color from the current palette.

Thanks, that's what I suspected, was just scratching my head as to where those CSS class definitions were being (or should be) generated.

Reply:
Thanks for the clarification. Just dropping here to add that I had the TwentyTwentyOne theme activated. 

I could see the expected behaviour after switching to TT1 Blocks.

Reply:
I've [added the missing filter](https://github.com/WordPress/gutenberg/pull/30124/commits/7a2a25c44198bb39e912928ca9c27f1e03a8bba8) to force the inline styles for themes that don't load their color palettes in the editor. 

Testing TwentyTwentyOne, the appropriate border colors are now correctly shown in the editor. The CSS classes on the frontend still do not appear to be having any effect there though.

Reply:
After a rebase, the named color css classes do work for the TwentyTwentyOne theme. TwentyTwenty and some other older themes don't look to be qualifying for the generation of the css classes via the theme.json. I'm still looking into those.

Reply:
@nosolosw would you be able to point me in the right direction for getting the color styles generated from the theme.json color palette available for themes such as TwentyTwenty?

I believe the active theme currently has to satisfy [this check](https://github.com/WordPress/gutenberg/blob/158b115556a7502da7dc0bd88ffe28ad37569a74/lib/global-styles.php#L136) before the necessary styles are included.

Reply:
`useEditorFeature( 'color.palette' )` will have the color palette, whether or not the theme has support for theme.json

Reply:
The issue I'm trying to solve here is some older themes don't qualify to have the named color CSS classes included. In other words they are missing `has-<color>-border-color`.

If the TT1 Blocks or TwentyTwenty Blocks themes are active, they get the generated color classes fine. Something older such as the TwentyTwenty (non block based) theme doesn't get these so while the block has the correct CSS class on the frontend the CSS class doesn't exist in the stylesheet.

Reply:
Ah, I see. Themes that don't support theme.json won't have the preset classes automatically enqueued for them, they add them themselves. At some point in the past, we experimented with enqueue the classes for these themes as well but heard reports that this wasn't expected, so we stop doing it.

What I see is that named border-color classes are properly attached to the block (for any theme), and that's the only thing relevant for the hook. In terms of the global stylesheet, I also see them generated and enqueued in the embedded stylesheet `global-styles-inline-css` for themes with theme.json support, so everything is working as expected.

One thing we should do is to add this as a dev note saying: there's this new property so themes need to enqueue the classes for the border-colors as well. Only if this feature becomes stable for 5.8. I'll add the label so we can find them later.

Also ccing a few theme people FYI (re: `has-<color>-border-color` classes) ― @carolinan @aristath @scruffian @MaggieCabrera

Reply:
Thanks for the explanation. That helps my understanding 👍